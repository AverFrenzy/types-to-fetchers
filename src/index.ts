import axios, { AxiosError } from 'axios';
import { compile } from 'path-to-regexp';

export type UnionToIntersection<U> = (
  U extends any ? (k: U) => void : never
) extends (k: infer I) => void
  ? I
  : never;

export type LastOf<T> = UnionToIntersection<
  T extends any ? () => T : never
> extends () => infer R
  ? R
  : never;

export type TuplifyUnion<
  T,
  L = LastOf<T>,
  N = [T] extends [never] ? true : false
> = true extends N ? [] : [...TuplifyUnion<Exclude<T, L>>, L];

export type Effect<Config extends BaseConfig, Error> = (
  action: Fetcher<Config, Error>
) => Fetcher<Config, Error>;

export type BaseConfig = {
  Body?: unknown;
  Querystring?: unknown;
  Params?: Record<string, string>;
  Headers?: unknown;
  Reply: unknown;
};

export type Options<Config extends BaseConfig, Error> = {
  baseURL: string;
  effect?: Effect<Config, Error>;
};

export type Input<API> = {
  [Endpoint in keyof API]: TuplifyUnion<keyof API[Endpoint]>;
};

export type Output<API> = {
  [Endpoint in keyof API]: {
    [Method in keyof API[Endpoint]]: Fetcher<
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      API[Endpoint][Method],
      Error
    >;
  };
};

export type Fetcher<Config extends BaseConfig, Error> = (
  data: Omit<Config, 'Reply' | 'Headers'>
) => Promise<Exclude<Config['Reply'], Error>>;

export const fetcher =
  <Config extends BaseConfig, Error>(
    baseURL: string,
    url: string,
    method: string
  ): Fetcher<Config, Error> =>
  async ({ Body, Querystring, Params }) => {
    try {
      const { data } = await axios({
        url: compile(url)(Params),
        method,
        baseURL,
        data: Body,
        params: Querystring,
        withCredentials: true,
      });

      return data;
    } catch (error) {
      const { response, message } = error as AxiosError<Error>;

      throw (response?.data as any)?.error ?? message ?? 'Unknown error';
    }
  };

export const makeApi = <API, Error, Effect = void>(
  schema: Input<API>,
  options: Options<BaseConfig, Error>
): Effect extends void ? Output<API> : Effect => {
  const result = schema as any;
  const { baseURL, effect } = options;

  for (const endpoint in result) {
    result[endpoint] = result[endpoint].reduce(
      (acc: Record<string, Fetcher<BaseConfig, Error>>, method: string) => {
        const handler = fetcher(baseURL, endpoint, method);
        acc[method] = effect ? effect(handler) : handler;

        return acc;
      },
      {}
    );
  }

  return result;
};
