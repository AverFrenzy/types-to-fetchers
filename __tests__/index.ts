import { fetcher, makeApi } from '../src';
import axios from 'axios';

jest.mock('axios');
const mocked = jest.mocked(axios, { shallow: true });
mocked.mockResolvedValue({ data: { foo: 'bar' } } as never);

interface API {
  '/': {
    GET: {
      Reply: {
        version: string;
        mode: 'production' | 'development';
      };
    };
  };
  '/foo/:bar': {
    GET: {
      Params: { bar: string };
      Reply: string;
    };
    POST: {
      Params: { bar: string };
      Body: { baz: string };
      Reply: Error | string;
    };
  };
}

describe('Fetcher', () => {
  test('Fetcher call returns a function', () => {
    expect(typeof fetcher('', '', '', new AbortController().signal)).toBe(
      'function'
    );
    expect(
      typeof fetcher(
        'https://example.com',
        '/',
        'GET',
        new AbortController().signal
      )
    ).toBe('function');
  });

  test("Url without format expression won't compile", () => {
    const abortController = new AbortController();

    fetcher('', '/x', 'GET', abortController.signal)({});

    expect(mocked).toHaveBeenCalledWith({
      baseURL: '',
      method: 'GET',
      url: '/x',
      withCredentials: true,
      signal: abortController.signal,
    });
  });

  test('Compiling a url with a format expression will return the compiled url', () => {
    const abortController = new AbortController();

    fetcher('', '/x/:y', 'GET', abortController.signal)({ Params: { y: 'z' } });

    expect(mocked).toHaveBeenCalledWith({
      baseURL: '',
      method: 'GET',
      url: '/x/z',
      withCredentials: true,
      signal: abortController.signal,
    });
  });

  test('The Body forwards correctly', () => {
    const abortController = new AbortController();

    fetcher('', '/', 'POST', abortController.signal)({ Body: { a: 'b' } });

    expect(mocked).toHaveBeenCalledWith({
      baseURL: '',
      method: 'POST',
      url: '/',
      withCredentials: true,
      data: { a: 'b' },
      signal: abortController.signal,
    });
  });

  test('The Querystring forwards correctly', () => {
    const abortController = new AbortController();

    fetcher(
      '',
      '/',
      'GET',
      abortController.signal
    )({ Querystring: { a: 'b' } });

    expect(mocked).toHaveBeenCalledWith({
      baseURL: '',
      method: 'GET',
      url: '/',
      withCredentials: true,
      params: { a: 'b' },
      signal: abortController.signal,
    });
  });

  test('Response returned correctly', async () => {
    const response = await fetcher(
      '',
      '/',
      'GET',
      new AbortController().signal
    )({});

    expect(response).toEqual({ foo: 'bar' });
  });

  test('Signal exist, work correctly', async () => {
    const controller = new AbortController();

    const response = await fetcher('', '/', 'GET', controller.signal)({});

    expect(axios).toHaveBeenCalledWith(
      expect.objectContaining({
        signal: controller.signal,
      })
    );

    expect(response).toEqual({ foo: 'bar' });
  });
});

describe('Make API', () => {
  test('The generator output snapshot is correctly', () => {
    const api = makeApi<API, {}>(
      {
        '/': ['GET'],
        '/foo/:bar': ['GET', 'POST'],
      },
      { baseURL: '/x' }
    );

    expect(api).toMatchSnapshot();
  });

  test('The effect forwards correctly', async () => {
    const api = makeApi<API, {}>(
      {
        '/': ['GET'],
        '/foo/:bar': ['GET', 'POST'],
      },
      {
        baseURL: 'hrrps://api.mysite.com/',
        effect: (_, params) => params as never,
      }
    );

    expect(api).toMatchSnapshot();
  });

  test('The baseURL forwards correctly', () => {
    const api = makeApi<API, {}>(
      {
        '/': ['GET'],
        '/foo/:bar': ['GET', 'POST'],
      },
      { baseURL: 'hrrps://api.mysite.com/' }
    );

    api['/'].GET({});

    expect(mocked).toHaveBeenCalledWith({
      baseURL: 'hrrps://api.mysite.com/',
      method: 'GET',
      url: '/',
      withCredentials: true,
      signal: expect.any(AbortSignal),
    });
  });

  test('The api has abort function', () => {
    const api = makeApi<API, {}>(
      {
        '/': ['GET'],
        '/foo/:bar': ['GET', 'POST'],
      },
      { baseURL: 'hrrps://api.mysite.com/' }
    );

    api['/'].GET({});
    api['/'].GET.abort();

    Object.values(api).forEach((endpoint) => {
      Object.values(endpoint).forEach((properties) => {
        console.log(properties);
        expect(properties).toHaveProperty('abort');
        expect(typeof properties.abort).toBe('function');
      });
    });
  });
});
