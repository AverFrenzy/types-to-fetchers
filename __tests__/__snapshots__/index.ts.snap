// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Make API The effect forwards correctly 1`] = `
{
  "/": {
    "GET": {
      "abort": [Function],
      "baseURL": "hrrps://api.mysite.com/",
      "endpoint": "/",
      "method": "GET",
    },
  },
  "/foo/:bar": {
    "GET": {
      "abort": [Function],
      "baseURL": "hrrps://api.mysite.com/",
      "endpoint": "/foo/:bar",
      "method": "GET",
    },
    "POST": {
      "abort": [Function],
      "baseURL": "hrrps://api.mysite.com/",
      "endpoint": "/foo/:bar",
      "method": "POST",
    },
  },
}
`;

exports[`Make API The generator output snapshot is correctly 1`] = `
{
  "/": {
    "GET": [Function],
  },
  "/foo/:bar": {
    "GET": [Function],
    "POST": [Function],
  },
}
`;
